File: Programmers/Level2/피로도/피로도.py
from itertools import permutations

def solution(k, dungeons):
    answer = []
    
    for perm in permutations(dungeons, len(dungeons)):
        count = 0
        tmp = k
        
        for p in perm:
            need, use = p
            if tmp < need:
                break    
            tmp -= use
            count += 1
        answer.append(count)
        
    return max(answer)


# dfs 이용 풀이
answer = 0

def dfs(k, count, dungeons, visited):
    global answer
    if count > answer: answer = count
    
    for i in range(len(dungeons)):
        if k >= dungeons[i][0] and not visited[i]:
            visited[i] = True
            dfs(k - dungeons[i][1], count + 1, dungeons, visited)
            visited[i] = False
        
    
def solution(k, dungeons):
    visited = [False] * len(dungeons)
    dfs(k, 0, dungeons, visited)
    return answer
    
    

File: Programmers/Level2/피로도/피로도.js
const getPermutations = (array, selectNumber) => {
    const results = [];
    if (selectNumber === 1) {
        return array.map((element) => [element]);
    }
    array.forEach((fixed, index, origin) => {
        const rest = [...origin.slice(0, index), ...origin.slice(index + 1)];
        const permutations = getPermutations(rest, selectNumber - 1);
        const attached = permutations.map((permutation) => [
            fixed,
            ...permutation,
        ]);

        results.push(...attached);
    });
    return results;
};

function solution(k, dungeons) {
    answer = [];
    for (let perm of getPermutations(dungeons, dungeons.length)) {
        let count = 0;
        let tmp = k;
        for (let p of perm) {
            let [need, use] = p;
            if (tmp < need) break;
            tmp -= use;
            count += 1;
        }
        answer.push(count);
    }
    return Math.max(...answer);
}
